File: domains/migration/go_engine/go_job_dispatch_config.py
"""
Module: go_job_dispatch_config.py
Purpose: Serializable job dispatch contract stored in migration_jobs.config JSONB.
Author: Ravi Sharma
Copyright (c) 2026 Ravi Sharma
SPDX-License-Identifier: MIT
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any
from uuid import UUID

from domains.migration.go_engine.go_connection_dispatch_ref import GoConnectionDispatchRef
from domains.migration.go_engine.go_executor_kind import GoExecutorKind
from domains.migration.go_engine.go_table_dispatch_payload import GoTableDispatchPayload
from domains.migration.source_throttle import SourceThrottleConfig


@dataclass(frozen=True, slots=True)
class GoJobDispatchConfig:
    """Control-plane → Go data-plane handoff document (migration_jobs.config)."""

    job_id: UUID
    executor: GoExecutorKind
    source: GoConnectionDispatchRef
    target: GoConnectionDispatchRef
    tables: tuple[GoTableDispatchPayload, ...]
    snapshot_ref: str | None = None
    idempotent: bool = False
    conflict_columns: tuple[str, ...] = field(default_factory=tuple)
    use_nolock: bool = False
    source_throttle: SourceThrottleConfig = field(default_factory=SourceThrottleConfig)

    def to_dict(self) -> dict[str, Any]:
        out: dict[str, Any] = {
            "job_id": str(self.job_id),
            "executor": self.executor.value,
            "source": self.source.to_dict(),
            "target": self.target.to_dict(),
            "tables": [t.to_dict() for t in self.tables],
            "snapshot_ref": self.snapshot_ref,
            "idempotent": self.idempotent,
            "use_nolock": self.use_nolock,
            "source_throttle": self.source_throttle.to_dict(),
        }
        if self.conflict_columns:
            out["conflict_columns"] = list(self.conflict_columns)
        return out

    @classmethod
    def from_dict(cls, data: dict[str, Any]) -> GoJobDispatchConfig:
        if data.get("executor") != GoExecutorKind.GO.value:
            raise ValueError(f"Unsupported executor in dispatch config: {data.get('executor')!r}")
        tables_raw = data.get("tables") or []
        conflict_raw = data.get("conflict_columns") or []
        return cls(
            job_id=UUID(str(data["job_id"])),
            executor=GoExecutorKind.GO,
            source=GoConnectionDispatchRef.from_dict(data["source"]),
            target=GoConnectionDispatchRef.from_dict(data["target"]),
            tables=tuple(GoTableDispatchPayload.from_dict(t) for t in tables_raw),
            snapshot_ref=data.get("snapshot_ref"),
            idempotent=bool(data.get("idempotent", False)),
            conflict_columns=tuple(str(c) for c in conflict_raw),
            use_nolock=bool(data.get("use_nolock", False)),
            source_throttle=SourceThrottleConfig.from_dict(data.get("source_throttle")),
        )

    def validate(self) -> None:
        """Raise ValueError when the contract is incomplete for the Go worker."""
        if not self.tables:
            raise ValueError("GoJobDispatchConfig requires at least one table")
        for table in self.tables:
            if not table.columns:
                raise ValueError(f"Table {table.table_name} has no resolved columns")
            if "*" in table.columns:
                raise ValueError(f"Table {table.table_name} must not use wildcard columns")

File: engine-go/internal/worker/go_job_dispatch_config.go
// Copyright (c) 2026 Ravi Sharma. MIT License.
// SPDX-License-Identifier: MIT

package worker

import (
	"encoding/json"
	"fmt"

	"github.com/google/uuid"
)

// GoJobDispatchConfig mirrors the Python control-plane JSON contract.
type GoJobDispatchConfig struct {
	JobID            string                    `json:"job_id"`
	Executor         string                    `json:"executor"`
	Source           GoConnectionDispatchRef   `json:"source"`
	Target           GoConnectionDispatchRef   `json:"target"`
	Tables           []GoTableDispatchPayload  `json:"tables"`
	SnapshotRef      *string                   `json:"snapshot_ref"`
	Idempotent       bool                      `json:"idempotent"`
	ConflictColumns  []string                  `json:"conflict_columns"`
	UseNoLock        bool                      `json:"use_nolock"`
	SourceThrottle   SourceThrottleConfig      `json:"source_throttle"`
}

type GoConnectionDispatchRef struct {
	ConnectionID string `json:"connection_id"`
	Schema       string `json:"schema"`
}

type GoTableDispatchPayload struct {
	TableName         string            `json:"table_name"`
	SourceSchema      string            `json:"source_schema"`
	TargetSchema      string            `json:"target_schema"`
	Columns           []string          `json:"columns"`
	ChunkSize         int               `json:"chunk_size"`
	ParallelWorkers   int               `json:"parallel_workers"`
	Strategy          string            `json:"strategy"`
	ColumnTransforms  map[string]string `json:"column_transforms"`
	ColumnSensitivity map[string]string `json:"column_sensitivity"`
	ColumnTypes        map[string]string `json:"column_types"`
	ColumnExtractCasts map[string]string `json:"column_extract_casts"`
	OrderColumn        string            `json:"order_column"`
	WhereClause       string            `json:"where_clause"`
	SourceMaxDOP      int               `json:"source_maxdop"`
	SkipDataLoad      bool              `json:"skip_data_load"`
	RowCountEstimate  int64             `json:"row_count_estimate"`
	TableSizeMB       float64           `json:"table_size_mb"`
	ChunkDelaySec     float64           `json:"chunk_delay_sec"`
}

// ParseGoJobDispatchConfig validates and parses raw JSON from migration_jobs.config.
func ParseGoJobDispatchConfig(raw json.RawMessage) (*GoJobDispatchConfig, error) {
	if len(raw) == 0 {
		return nil, fmt.Errorf("empty dispatch config")
	}
	var cfg GoJobDispatchConfig
	if err := json.Unmarshal(raw, &cfg); err != nil {
		return nil, fmt.Errorf("unmarshal dispatch config: %w", err)
	}
	if cfg.Executor != "go" {
		return nil, fmt.Errorf("unsupported executor %q", cfg.Executor)
	}
	if len(cfg.Tables) == 0 {
		return nil, fmt.Errorf("dispatch config has no tables")
	}
	for _, t := range cfg.Tables {
		if len(t.Columns) == 0 {
			return nil, fmt.Errorf("table %s has no columns", t.TableName)
		}
	}
	if _, err := uuid.Parse(cfg.JobID); err != nil {
		return nil, fmt.Errorf("invalid job_id: %w", err)
	}
	return &cfg, nil
}
